PageSpeed_Api.php

```diff
@@ -1,128 +1,600 @@
+			return wp_json_encode(
+				array(
+					'error' => array(
+						'code'    => $response->get_error_code(),
+						'message' => $response->get_error_message(),
+					),
+				)
+			);
+		} elseif ( isset( $response['error']['code'] ) && 200 !== $response['error']['code'] ) {
+			if ( 'get-token-missing-site-id' === $response['error']['id'] ) {
+				$message = __( 'No site ID provided for Google access record update!', 'w3-total-cache' );
+			} elseif ( 'get-token-missing-w3tc-pagespeed-key' === $response['error']['id'] ) {
+				$message = __( 'No W3 key provided for Google access record update!', 'w3-total-cache' );
+			} elseif ( 'get-token-not-found' === $response['error']['id'] ) {
+				$message = __( 'No matching Google access record found for W3 key!', 'w3-total-cache' );
+			} elseif ( 'get-token-bad-record' === $response['error']['id'] ) {
+				$message = __( 'Matching Google access record found but the refresh token value is blank!', 'w3-total-cache' );
+			}
+
+			return wp_json_encode(
+				array(
+					'error' => array(
+						'code'    => $response['error']['code'],
+						'message' => $message,
+					),
+				)
+			);
 		}
 
-		$key = array_shift( $elements );
-		if ( !isset( $data[$key] ) ) {
-			return null;
+		// Response body should contain a JSON format string.
+		return wp_remote_retrieve_body( $response );
+	}
+
+	/**
+	 * Get Google Client JSON config.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @return string
+	 */
+	public function get_client_json() {
+		$client_json = defined( 'W3TC_GOOGLE_CLIENT_JSON' ) && W3TC_GOOGLE_CLIENT_JSON ? W3TC_GOOGLE_CLIENT_JSON : $this->google_client_json;
+		$client      = json_decode( $client_json );
+		foreach ( $client->web->redirect_uris as $redirect_uri_key => $redirect_uri_value ) {
+			$client->web->redirect_uris[ $redirect_uri_key ] = $this->get_w3tc_api_url( $redirect_uri_value );
 		}
+		return wp_json_encode( $client );
+	}
+
+	/**
+	 * Get W3TC PageSpeed API max attempts.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @return int
+	 */
+	public function get_max_attempts() {
+		return defined( 'W3TC_PAGESPEED_MAX_ATTEMPTS' ) && W3TC_PAGESPEED_MAX_ATTEMPTS ? W3TC_PAGESPEED_MAX_ATTEMPTS : $this->retry_attempts;
+	}
+
+	/**
+	 * Get Google PageSpeed API URL.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @return string
+	 */
+	public function get_pagespeed_url() {
+		return defined( 'W3TC_PAGESPEED_API_URL' ) && W3TC_PAGESPEED_API_URL ? W3TC_PAGESPEED_API_URL : $this->pagespeed_api_base_url;
+	}
+
+	/**
+	 * Get W3TC API server URL target. Will be forced to HTTPS.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string $target API target URI.
+	 *
+	 * @return string
+	 */
+	public function get_w3tc_api_url( $target ) {
+		// ONLY FOR BETA - REMOVE BEFORE PRODUCTION.
+		define( 'W3TC_API2_URL', 'api2-dev-jacobd.w3-edge.com' );
+
+		return defined( 'W3TC_API2_URL' ) && W3TC_API2_URL ? esc_url( 'https://' . W3TC_API2_URL . '/' . $target, array( 'https' ), '' ) : esc_url( $this->w3tc_api_base_url . '/' . $target, 'https', '' );
```

Comment:
I've removed the esc_url call here and updated it slightly to use trailingslashit to prevent double //. The use of esc_url is only needed on display or print-to-screen

Additionally the 2nd parameter of esc_url actually expects an array of strings but is optional and will default to null if omitted